File: client/go/conn.go
package teraslab

import (
	"context"
	"fmt"
	"net"
	"sync"
	"sync/atomic"
	"time"
)

// respChanPool pools the per-request response channels to avoid
// allocating a new channel on every roundTrip call.
var respChanPool = sync.Pool{
	New: func() any { return make(chan responseFrame, 1) },
}

// testHookReadLoopDeliver, when non-nil, is invoked by readLoop after it has
// claimed a pending channel (LoadAndDelete) but before it sends the response on
// that channel. It is nil in production (a single nil check per response) and
// exists only so tests can deterministically interleave a caller's
// cancellation with response delivery.
var testHookReadLoopDeliver func(reqID uint64)

// pipeConn is a pipelined TCP connection that supports multiple in-flight
// requests matched by request_id. It is safe for concurrent use.
type pipeConn struct {
	conn      net.Conn
	mu        sync.Mutex // serializes writes
	writeBuf  []byte     // reusable write buffer, protected by mu
	pending   sync.Map   // map[uint64]chan responseFrame
	nextID    atomic.Uint64
	inflight  atomic.Int64 // count of in-flight roundTrips (incl. health pings)
	readDone  chan struct{}
	closeOnce sync.Once
	closed    atomic.Bool
	connErr   atomic.Pointer[error]
}

// hasInflight reports whether the connection currently has in-flight requests.
// The pool uses this to avoid health-probing or reaping a connection that is in
// active use: a pipelined conn carrying live requests does not need a liveness
// ping (a genuine failure surfaces via the read/write path), and closing it
// would abort those legitimate in-flight requests with "connection closed".
func (c *pipeConn) hasInflight() bool {
	return c.inflight.Load() > 0
}

// inflightCount returns the number of requests currently in flight on this
// connection (including health pings). The pool uses this to pick the
// least-loaded connection and to decide whether a connection still has headroom
// below the pipeline-depth target before growing the pool.
func (c *pipeConn) inflightCount() int64 {
	return c.inflight.Load()
}

// dial creates a new pipelined connection to the given address.
func dial(ctx context.Context, addr string, timeout time.Duration) (*pipeConn, error) {
	d := net.Dialer{Timeout: timeout}
	conn, err := d.DialContext(ctx, "tcp", addr)
	if err != nil {
		return nil, fmt.Errorf("dial %s: %w", addr, err)
	}
	// Disable Nagle's algorithm. Requests and responses are small framed
	// messages; with Nagle on, the client's small request writes are held in
	// the kernel TCP send buffer waiting for a peer ACK, interacting with the
	// peer's delayed-ACK to add 40 ms–3 s of latency per RPC. The server sets
	// TCP_NODELAY on its accept path (src/server/mod.rs); without the matching
	// client-side setting, every round-trip still pays the Nagle/delayed-ACK
	// tax. Best-effort: only *net.TCPConn supports it.
	if tcp, ok := conn.(*net.TCPConn); ok {
		_ = tcp.SetNoDelay(true)
	}
	pc := &pipeConn{
		conn:     conn,
		writeBuf: make([]byte, 0, 4096),
		readDone: make(chan struct{}),
	}
	go pc.readLoop()
	return pc, nil
}

// releaseChan returns ch to respChanPool, but only if this goroutine still owns
// the pending entry for reqID. Winning the LoadAndDelete guarantees that neither
// readLoop nor closeInternal will send on ch (they remove the entry before
// sending), so ch is empty and safe to reuse. If we lose the race, the winner
// will send a frame on ch; returning it to the pool would leak that buffered
// frame into the next request that reuses the channel, so we drop it instead and
// let the GC reclaim it.
func (c *pipeConn) releaseChan(reqID uint64, ch chan responseFrame) {
	if _, ok := c.pending.LoadAndDelete(reqID); ok {
		respChanPool.Put(ch)
	}
}

// roundTrip sends a request and waits for its response.
// The request_id is assigned automatically.
func (c *pipeConn) roundTrip(ctx context.Context, opCode uint16, flags uint16, payload []byte) (responseFrame, error) {
	if c.closed.Load() {
		return responseFrame{}, c.getErr()
	}

	reqID := c.nextID.Add(1)
	ch := respChanPool.Get().(chan responseFrame)
	c.pending.Store(reqID, ch)
	c.inflight.Add(1)
	defer c.inflight.Add(-1)

	// Write the request using the connection's reusable write buffer.
	f := &requestFrame{
		RequestID: reqID,
		OpCode:    opCode,
		Flags:     flags,
		Payload:   payload,
	}

	c.mu.Lock()
	var err error
	c.writeBuf, err = writeRequest(c.conn, f, c.writeBuf)
	c.mu.Unlock()
	if err != nil {
		c.releaseChan(reqID, ch)
		return responseFrame{}, fmt.Errorf("write: %w", err)
	}

	// Wait for response or context cancellation.
	select {
	case resp := <-ch:
		respChanPool.Put(ch)
		// Zero RequestID with closed connection means we were woken by closeInternal.
		if resp.RequestID == 0 && c.closed.Load() {
			return responseFrame{}, c.getErr()
		}
		return resp, nil
	case <-ctx.Done():
		c.releaseChan(reqID, ch)
		return responseFrame{}, ctx.Err()
	case <-c.readDone:
		c.releaseChan(reqID, ch)
		return responseFrame{}, c.getErr()
	}
}

// roundTripSigned is like roundTrip but HMAC-signs the WHOLE frame body
// (request_id || op_code || flags || payload) with the cluster secret, as
// required by the server's strict_auth gate for inter-node opcodes (e.g.
// OP_GET_PARTITION_MAP). The frame is signed AFTER request_id is assigned via
// signFrame (mirroring src/cluster/auth.rs::sign_frame) so the signature covers
// the exact bytes on the wire.
func (c *pipeConn) roundTripSigned(ctx context.Context, opCode uint16, flags uint16, payload, secret []byte) (responseFrame, error) {
	if c.closed.Load() {
		return responseFrame{}, c.getErr()
	}

	reqID := c.nextID.Add(1)
	ch := respChanPool.Get().(chan responseFrame)
	c.pending.Store(reqID, ch)
	c.inflight.Add(1)
	defer c.inflight.Add(-1)

	f := &requestFrame{
		RequestID: reqID,
		OpCode:    opCode,
		Flags:     flags,
		Payload:   payload,
	}
	encoded := encodeRequest(nil, f)
	signed := signFrame(secret, encoded, uint64(time.Now().UnixMilli()))

	c.mu.Lock()
	_, err := c.conn.Write(signed)
	c.mu.Unlock()
	if err != nil {
		c.releaseChan(reqID, ch)
		return responseFrame{}, fmt.Errorf("write: %w", err)
	}

	select {
	case resp := <-ch:
		respChanPool.Put(ch)
		if resp.RequestID == 0 && c.closed.Load() {
			return responseFrame{}, c.getErr()
		}
		return resp, nil
	case <-ctx.Done():
		c.releaseChan(reqID, ch)
		return responseFrame{}, ctx.Err()
	case <-c.readDone:
		c.releaseChan(reqID, ch)
		return responseFrame{}, c.getErr()
	}
}

// readLoop runs in a goroutine, reading response frames and dispatching
// them to waiting callers.
func (c *pipeConn) readLoop() {
	defer close(c.readDone)
	var buf []byte
	for {
		resp, newBuf, err := readResponse(c.conn, buf)
		buf = newBuf
		if err != nil {
			c.setErr(fmt.Errorf("read: %w", err))
			c.closeInternal()
			return
		}
		if ch, ok := c.pending.LoadAndDelete(resp.RequestID); ok {
			if testHookReadLoopDeliver != nil {
				testHookReadLoopDeliver(resp.RequestID)
			}
			ch.(chan responseFrame) <- resp
		} else {
			// No pending entry — response is for a cancelled request.
			// Recycle the payload since nobody will consume it.
			recyclePayload(resp.Payload)
		}
	}
}

// close closes the connection and wakes all pending callers.
func (c *pipeConn) close() error {
	c.setErr(fmt.Errorf("connection closed"))
	return c.closeInternal()
}

func (c *pipeConn) closeInternal() error {
	var closeErr error
	c.closeOnce.Do(func() {
		c.closed.Store(true)
		closeErr = c.conn.Close()
		// Wake all pending callers by sending a zero-value response on their
		// channels, then return channels to the pool. We don't close pooled
		// channels because they may be reused.
		c.pending.Range(func(key, value any) bool {
			ch := value.(chan responseFrame)
			select {
			case ch <- responseFrame{}:
			default:
			}
			c.pending.Delete(key)
			return true
		})
	})
	return closeErr
}

// alive returns true if the connection is healthy.
func (c *pipeConn) alive() bool {
	return !c.closed.Load()
}

func (c *pipeConn) setErr(err error) {
	c.connErr.CompareAndSwap(nil, &err)
}

func (c *pipeConn) getErr() error {
	if p := c.connErr.Load(); p != nil {
		return *p
	}
	return fmt.Errorf("connection closed")
}

File: client/rust/src/conn.rs
//! Pipelined TCP connection for the TeraSlab wire protocol.
//!
//! Each [`PipeConn`] multiplexes many in-flight requests over a single TCP
//! connection, matching responses to callers by `request_id`. This mirrors the
//! Go client's `pipeConn`.

use std::collections::HashMap;
use std::sync::Arc;
use std::sync::atomic::{AtomicBool, AtomicU64, Ordering};
use std::time::Duration;

use parking_lot::Mutex;
use teraslab::protocol::frame::{RequestFrame, ResponseFrame};
use teraslab::protocol::opcodes::MAX_FRAME_SIZE;
use tokio::io::{AsyncReadExt, AsyncWriteExt};
use tokio::net::TcpStream;
use tokio::sync::oneshot;
use tokio::task::JoinHandle;

use crate::errors::ClientError;

/// A pipelined TCP connection that supports multiple concurrent in-flight
/// requests matched by `request_id`.
///
/// The connection runs a background read loop (spawned Tokio task) that
/// dispatches incoming response frames to the callers waiting on oneshot
/// channels.
pub(crate) struct PipeConn {
    /// The write half, protected by a mutex to serialize writes.
    writer: tokio::sync::Mutex<tokio::io::WriteHalf<TcpStream>>,
    /// Map of pending request_id -> oneshot sender for response delivery.
    pending: Arc<Mutex<HashMap<u64, oneshot::Sender<ResponseFrame>>>>,
    /// Atomically incrementing request ID counter.
    next_id: AtomicU64,
    /// Whether this connection is still alive, shared with the read loop.
    alive: Arc<AtomicBool>,
    /// Handle to the background read loop task. Kept alive until dropped.
    _read_task: JoinHandle<()>,
}

impl PipeConn {
    /// Establish a new pipelined connection to the given address.
    ///
    /// # Errors
    ///
    /// Returns [`ClientError::Connection`] if the TCP connection fails or times out.
    pub async fn dial(addr: &str, timeout: Duration) -> Result<Self, ClientError> {
        let stream = tokio::time::timeout(timeout, TcpStream::connect(addr))
            .await
            .map_err(|_| ClientError::Connection(format!("dial timeout: {}", addr)))?
            .map_err(|e| ClientError::Connection(format!("dial {}: {}", addr, e)))?;

        // Disable Nagle's algorithm for lower latency.
        stream
            .set_nodelay(true)
            .map_err(|e| ClientError::Connection(format!("set_nodelay: {}", e)))?;

        let (reader, writer) = tokio::io::split(stream);

        let pending: Arc<Mutex<HashMap<u64, oneshot::Sender<ResponseFrame>>>> =
            Arc::new(Mutex::new(HashMap::new()));
        let alive = Arc::new(AtomicBool::new(true));

        let read_pending = Arc::clone(&pending);
        let read_alive = Arc::clone(&alive);

        let read_task = tokio::spawn(async move {
            read_loop(reader, read_pending, read_alive).await;
        });

        Ok(Self {
            writer: tokio::sync::Mutex::new(writer),
            pending,
            next_id: AtomicU64::new(0),
            alive,
            _read_task: read_task,
        })
    }

    /// Send a request and wait for its response.
    ///
    /// The `request_id` is assigned automatically. The method serializes the
    /// request frame, writes it under the write lock, and awaits the response
    /// on a oneshot channel.
    ///
    /// # Errors
    ///
    /// Returns [`ClientError::Connection`] if the connection is dead or the
    /// write fails, or [`ClientError::Timeout`] if the default timeout elapses.
    pub async fn round_trip(
        &self,
        op_code: u16,
        flags: u16,
        payload: Vec<u8>,
    ) -> Result<ResponseFrame, ClientError> {
        if !self.alive.load(Ordering::Acquire) {
            return Err(ClientError::Connection("connection closed".to_string()));
        }

        let req_id = self.next_id.fetch_add(1, Ordering::Relaxed) + 1;
        let (tx, rx) = oneshot::channel();

        // Register the pending request.
        {
            let mut pending = self.pending.lock();
            pending.insert(req_id, tx);
        }

        // Encode and write the request frame.
        let frame = RequestFrame {
            request_id: req_id,
            op_code,
            flags,
            payload,
        };
        let encoded = frame.encode();

        {
            let mut writer = self.writer.lock().await;
            if let Err(e) = writer.write_all(&encoded).await {
                self.pending.lock().remove(&req_id);
                return Err(ClientError::Connection(format!("write: {}", e)));
            }
        }

        // Wait for the response with a generous timeout.
        match tokio::time::timeout(Duration::from_secs(30), rx).await {
            Ok(Ok(resp)) => Ok(resp),
            Ok(Err(_)) => {
                // Sender was dropped (connection closed).
                Err(ClientError::Connection("connection closed".to_string()))
            }
            Err(_) => {
                self.pending.lock().remove(&req_id);
                Err(ClientError::Timeout)
            }
        }
    }

    /// Returns true if the connection is still alive.
    pub fn alive(&self) -> bool {
        self.alive.load(Ordering::Acquire)
    }

    /// Close this connection, waking all pending callers.
    pub async fn close(&self) {
        self.alive.store(false, Ordering::Release);
        // Drop all pending senders to wake waiters.
        let mut pending = self.pending.lock();
        pending.clear();
    }
}

impl Drop for PipeConn {
    fn drop(&mut self) {
        self.alive.store(false, Ordering::Release);
        self._read_task.abort();
    }
}

/// Background read loop that reads response frames from the TCP stream
/// and dispatches them to the corresponding pending oneshot senders.
async fn read_loop(
    mut reader: tokio::io::ReadHalf<TcpStream>,
    pending: Arc<Mutex<HashMap<u64, oneshot::Sender<ResponseFrame>>>>,
    alive: Arc<AtomicBool>,
) {
    let mut len_buf = [0u8; 4];
    loop {
        // Read 4-byte length prefix.
        if reader.read_exact(&mut len_buf).await.is_err() {
            break;
        }
        let total_length = u32::from_le_bytes(len_buf);
        if !(9..=MAX_FRAME_SIZE).contains(&total_length) {
            break;
        }

        // Read the body.
        let body_len = total_length as usize;
        let mut body = vec![0u8; body_len];
        if reader.read_exact(&mut body).await.is_err() {
            break;
        }

        // Decode: request_id(8) + status(1) + payload
        if body.len() < 9 {
            break;
        }
        let request_id = u64::from_le_bytes(body[0..8].try_into().unwrap());
        let status = body[8];
        let payload = body[9..].to_vec();

        let resp = ResponseFrame {
            request_id,
            status,
            payload,
        };

        // Look up and dispatch.
        let sender = {
            let mut map = pending.lock();
            map.remove(&request_id)
        };
        if let Some(tx) = sender {
            // Ignore send errors (caller may have timed out and dropped the receiver).
            let _ = tx.send(resp);
        }
    }

    // Connection is dead. Mark alive as false and wake all pending callers.
    alive.store(false, Ordering::Release);

    // Drop all pending senders so callers get an error.
    let mut map = pending.lock();
    map.clear();
}
